File: src/phorge/api/endpoints/deployments.py
"""Deployment API endpoints."""

from __future__ import annotations

from phorge.api.client import ForgeClient
from phorge.api.models import Deployment


class DeploymentsAPI:
    def __init__(self, client: ForgeClient) -> None:
        self._client = client

    async def list(self, server_id: int, site_id: int) -> list[Deployment]:
        data = await self._client.get(
            f"/servers/{server_id}/sites/{site_id}/deployment-history"
        )
        return [Deployment(**d) for d in data.get("deployments", [])]

    async def get(self, server_id: int, site_id: int, deployment_id: int) -> Deployment:
        data = await self._client.get(
            f"/servers/{server_id}/sites/{site_id}/deployment-history/{deployment_id}"
        )
        return Deployment(**data["deployment"])

    async def get_output(self, server_id: int, site_id: int, deployment_id: int) -> str:
        data = await self._client.get(
            f"/servers/{server_id}/sites/{site_id}/deployment-history/{deployment_id}/output"
        )
        return data.get("output", "")

    async def deploy(self, server_id: int, site_id: int) -> None:
        await self._client.post(
            f"/servers/{server_id}/sites/{site_id}/deployment/deploy"
        )

    async def get_log(self, server_id: int, site_id: int) -> str:
        data = await self._client.get(
            f"/servers/{server_id}/sites/{site_id}/deployment/log"
        )
        return data.get("output", "")

    async def get_script(self, server_id: int, site_id: int) -> str:
        return await self._client.get_text(
            f"/servers/{server_id}/sites/{site_id}/deployment/script"
        )

    async def update_script(self, server_id: int, site_id: int, content: str) -> None:
        await self._client.put(
            f"/servers/{server_id}/sites/{site_id}/deployment/script",
            json={"content": content},
        )

    async def enable_quick_deploy(self, server_id: int, site_id: int) -> None:
        await self._client.post(f"/servers/{server_id}/sites/{site_id}/deployment")

    async def disable_quick_deploy(self, server_id: int, site_id: int) -> None:
        await self._client.delete(f"/servers/{server_id}/sites/{site_id}/deployment")

    async def reset_status(self, server_id: int, site_id: int) -> None:
        await self._client.post(
            f"/servers/{server_id}/sites/{site_id}/deployment/reset"
        )

File: internal/forge/deployments.go
package forge

import (
	"context"
	"fmt"
	"net/http"
)

type deploymentOutput struct {
	Output string `json:"output"`
}

type deploymentScript struct {
	Content    string `json:"content"`
	AutoSource bool   `json:"auto_source"`
}

func (s *DeploymentsService) sitePath(serverID, siteID int64, suffix string) string {
	return s.client.orgPath("/servers/%d/sites/%d/deployments%s", serverID, siteID, suffix)
}

// List returns deployment history for a site.
func (s *DeploymentsService) List(ctx context.Context, serverID, siteID int64) ([]Deployment, error) {
	return listResources[Deployment](s.client, ctx, s.sitePath(serverID, siteID, ""))
}

// Get returns a single deployment by ID.
func (s *DeploymentsService) Get(ctx context.Context, serverID, siteID, deployID int64) (*Deployment, error) {
	return getResource[Deployment](s.client, ctx, s.sitePath(serverID, siteID, fmt.Sprintf("/%d", deployID)))
}

// GetOutput returns the output of a specific deployment.
func (s *DeploymentsService) GetOutput(ctx context.Context, serverID, siteID, deployID int64) (string, error) {
	out, err := getResource[deploymentOutput](s.client, ctx, s.sitePath(serverID, siteID, fmt.Sprintf("/%d/log", deployID)))
	if err != nil {
		return "", err
	}
	return out.Output, nil
}

// Deploy triggers a new deployment for the site.
func (s *DeploymentsService) Deploy(ctx context.Context, serverID, siteID int64) error {
	return s.client.do(ctx, http.MethodPost, s.sitePath(serverID, siteID, ""), nil, nil)
}

// GetLog returns the latest deployment log for the site. The new API has no
// separate "live log" endpoint, so this fetches the most recent deployment
// via List and then its output via GetOutput.
func (s *DeploymentsService) GetLog(ctx context.Context, serverID, siteID int64) (string, error) {
	deployments, err := s.List(ctx, serverID, siteID)
	if err != nil {
		return "", err
	}
	if len(deployments) == 0 {
		return "", nil
	}
	return s.GetOutput(ctx, serverID, siteID, deployments[0].ID)
}

// GetScript returns the deployment script contents.
func (s *DeploymentsService) GetScript(ctx context.Context, serverID, siteID int64) (string, error) {
	script, err := getResource[deploymentScript](s.client, ctx, s.sitePath(serverID, siteID, "/script"))
	if err != nil {
		return "", err
	}
	return script.Content, nil
}

// UpdateScript replaces the deployment script content.
func (s *DeploymentsService) UpdateScript(ctx context.Context, serverID, siteID int64, content string) error {
	path := s.sitePath(serverID, siteID, "/script")
	body := map[string]string{"content": content}
	return s.client.do(ctx, http.MethodPut, path, body, nil)
}

// EnableQuickDeploy enables quick deploy (push-to-deploy) for the site.
func (s *DeploymentsService) EnableQuickDeploy(ctx context.Context, serverID, siteID int64) error {
	return s.client.do(ctx, http.MethodPost, s.sitePath(serverID, siteID, "/push-to-deploy"), nil, nil)
}

// DisableQuickDeploy disables quick deploy for the site.
func (s *DeploymentsService) DisableQuickDeploy(ctx context.Context, serverID, siteID int64) error {
	return s.client.do(ctx, http.MethodDelete, s.sitePath(serverID, siteID, "/push-to-deploy"), nil, nil)
}

// ResetStatus resets the deployment status for the site.
func (s *DeploymentsService) ResetStatus(ctx context.Context, serverID, siteID int64) error {
	return s.client.do(ctx, http.MethodDelete, s.sitePath(serverID, siteID, "/status"), nil, nil)
}
